linkage: display executables with missing rpath

chris-araman opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

An executable that links against `@rpath`-prefixed dylibs must include at least one `LC_RPATH`. This will prevent issues like the one resolved in https://github.com/Homebrew/homebrew-core/pull/91485.

Caveats:
1. This won't find executables that have only recursive dylib dependencies with @rpath prefixes.
1. This makes no attempt to resolve `@rpath`, `@executable_path` or `@loader_path` dylibs, or to verify the correctness of any `LC_RPATH` entries, or to otherwise simulate dlopen logic.
1. Weakly-linked dylibs are still excluded from the search for broken linkage.
1. I am not fluent in Ruby. 😊

The scope is narrow in order to focus on this particular problem. It is meant only as a sanity check.

chris-araman commented:
I haven't yet added new tests because I'm not yet sure how mock `ruby-macho` or how I might run `brew linkage` against known-bad and known-good binaries. Are old bottles stored indefinitely? Could I test against a known-bad `adig` binary, for instance?

carlocab commented on Library/Homebrew/linkage_checker.rb:
```diff
@@ -191,6 +194,12 @@ def check_dylibs(rebuild_cache:)
+      @executables_missing_rpaths << file if
+        Pathname(file).binary_executable? &&
+        dylibs.any? { |dylib| dylib.start_with? "@rpath/" } &&
+        MachO.open(file).rpaths.empty?
```
`MachO` is macOS-only, and I think we only want generic code here. 

We should have a generic `Pathname#rpaths` method you can use instead. This will also avoid the duplicate `MachO.open`, which is pretty expensive.

chris-araman replied:
Done.

carlocab commented on Library/Homebrew/linkage_checker.rb:
```diff
@@ -191,6 +194,12 @@ def check_dylibs(rebuild_cache:)
       end
+
+      # Binary executables that link @rpath-prefixed dylibs must include at least one LC_RPATH.
+      @executables_missing_rpaths << file if
```
Can we move this up into the block for handling `@variable_dylibs`?

chris-araman replied:
Done.

carlocab reviewed (commented):
Nice approach. Just a few comments.

carlocab commented:
> I haven't yet added new tests because I'm not yet sure how mock `ruby-macho` or how I might run `brew linkage` against known-bad and known-good binaries. Are old bottles stored indefinitely? Could I test against a known-bad `adig` binary, for instance?

We don't really have tests here precisely because of the lack of fixtures. This code is pretty heavily exercised in homebrew-core, though, so I wouldn't worry too much about tests as long as the code changes are straightforward (which they are here).

chris-araman commented:
I've addressed your feedback, @carlocab.

How can I ensure we're able to find and resolve the issue in any other formulae before merging this change?

carlocab commented on Library/Homebrew/linkage_checker.rb:
```diff
@@ -152,9 +155,18 @@ def check_dylibs(rebuild_cache:)
 
+        # Binary executables that link @rpath-prefixed dylibs must include at
+        # least one rpath in order to resolve it.
+        if !file_has_any_rpath_dylibs && (dylib.start_with? "@rpath/")
+          file_has_any_rpath_dylibs = true
+          pathname = Pathname(file)
+          @executables_missing_rpaths << file if pathname.binary_executable? && pathname.rpaths.empty?
+        end
```
Why not include this in the `if dylib.start_with? "@"` block?

chris-araman replied:
We want to find the full set of binaries that are missing rpaths. Moving this to that block would only examine the first binary to reference a given dylib.

carlocab commented on Library/Homebrew/linkage_checker.rb:
```diff
@@ -152,9 +155,18 @@ def check_dylibs(rebuild_cache:)
+        if !file_has_any_rpath_dylibs && (dylib.start_with? "@rpath/")
+          file_has_any_rpath_dylibs = true
+          pathname = Pathname(file)
+          @executables_missing_rpaths << file if pathname.binary_executable? && pathname.rpaths.empty?
```
It might be good to not restrict this to executables, since this can be an issue for libraries too, which can turn into an issue for executables.

For example, executable `foo` links with `/path/to/libbar.dylib`, but `libbar.dylib` links with `@rpath/libbaz.dylib`. If `libbar` does not have an `LC_RPATH` entry for `libbaz` then `foo` is also broken.

chris-araman replied:
That's an interesting case I hadn't considered.

Opening this check up to all binaries found that `mpdecimal`'s `libmpdec++.2.5.1.dylib` references `@rpath/libmpdec.3.dylib` but does not specify an `LC_RPATH`. In this case, it would still be possible for the `LC_RPATH` of the executable or another dependent dylib to resolve this.

Perhaps it would be better to walk the tree of backrefs until we find an `LC_RPATH`?

carlocab replied:
> In this case, it would still be possible for the LC_RPATH of the executable or another dependent dylib to resolve this.

Is it? I would find it surprising if Apple allows another binary's `LC_RPATH` entry to affect how the linker resolves library references. I tried testing it out locally too, and I don't think this is the case.

chris-araman replied:
My understanding is that the run-path search paths specified by executables are meant to resolve run-path dependent libraries loaded into the process.
https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/DynamicLibraries/100-Articles/RunpathDependentLibraries.html

What isn't spelled out in the documentation is whether the relationship between run-path search path and run-path dependent library only applies for direct library references, or whether it applies to recursively referenced libraries as well. That said, the documentation makes no mention of a _library_ specifying a run-path search path for run-path dependent libraries it references.

Let me do some testing and see what works and what doesn't, at least as of macOS 12.

chris-araman replied:
When executable `a` links `@rpath\libb.dylib` which links `@rpath\libc.dylib`, a search path specified in either `a` or `libb.dylib` can be used to resolve `@rpath\libc.dylib`.

Relative search paths are evaluated relative to the path of the executable, even when specified in a dylib found in a different path. For example, if executable `.\a` links `@rpath\libb.dylib` specifying search path `foo`, and `foo\libb.dylib` links `@rpath\libc.dylib` specifying search path `bar`, `libc.dylib` must be in `foo\` or `bar` (relative `a`) to be resolved. It will not be resolved from `foo\bar` (`bar` relative to `foo\libb.dylib`).

Given that a dylib can legitimately reference run-path dependent libraries without specifying a search path, letting the executable's search path resolve the reference, I think that we can really only perform this check on executables.

carlocab replied:
I'm not sure -- I don't think we want to ship libraries that contain `@rpath` references that rely on the user knowing to pass the right `-Wl,-rpath` flags to link the executables against the library. It would probably make linker failures easier to debug too.

chris-araman replied:
I agree that makes sense. I think it's clear enough that it's a Homebrew requirement and not an OS requirement. I've expanded the check to include executables, dylibs, and bundles.

carlocab reviewed (commented):
Apologies for the delay here -- got distracted by the holiday. This is looking good; just a few questions/suggestions.

chris-araman commented:
No worries. I'll come back to this next week.

carlocab reviewed (approved):
LGTM. Thanks, @chris-araman!

scpeters commented:
this has broken several of our formulae / bottles in the `osrf/simulation` tap. I'm switching our CI machines to use the latest brew stable tag instead of `master` for now, and will try to fix our bottles before next week's release. I may submit a pull request to make this behavior opt-in via an environment variable if I need more time

carlocab commented:
> I may submit a pull request to make this behavior opt-in via an environment variable if I need more time

We could move it to the `--strict` flag to make it a warning in `test-bot`. I would like to make this a hard failure in the long-term, however.

scpeters commented:
> > I may submit a pull request to make this behavior opt-in via an environment variable if I need more time
> 
> We could move it to the `--strict` flag to make it a warning in `test-bot`. I would like to make this a hard failure in the long-term, however.
> 
> Happy to help you try to fix any broken formulae resulting from this.

I've just set `HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK` for now, which is enough of a workaround

MikeMcQuaid commented:
> > Happy to help you try to fix any broken formulae resulting from this.
> 
> I've just set `HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK` for now, which is enough of a workaround

Just to check on this @scpeters: is this causing dependents to be reinstalled on install/upgrade due to this change? If so, that seems undesirable/unintentional?

carlocab commented:
> is this causing dependents to be reinstalled on install/upgrade due to this change?

Yes, I think this change would cause that.

> If so, that seems undesirable/unintentional?

Agreed. #12979

scpeters commented:
> > > Happy to help you try to fix any broken formulae resulting from this.
> > 
> > 
> > I've just set `HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK` for now, which is enough of a workaround
> 
> Just to check on this @scpeters: is this causing dependents to be reinstalled on install/upgrade due to this change? If so, that seems undesirable/unintentional?

yeah, it was causing them to try to reinstall, but I'm not sure it helped because a reinstall wouldn't fix the rpath
